packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java

```diff
@@ -0,0 +1,159 @@
+      promise.resolve((double)info.firstInstallTime);
+    } catch (Exception e) {
+      e.printStackTrace();
+      promise.reject("ERR_APPLICATION", "Unable to get first install time of this application.", e);
+    }
+  }
+
+  @ExpoMethod
+  public void getLastUpdateTimeAsync(Promise promise) {
+    PackageManager packageManager = mContext.getPackageManager();
+    String packageName = mContext.getPackageName();
+    try {
+      PackageInfo info = packageManager.getPackageInfo(packageName, 0);
+      promise.resolve((double)info.lastUpdateTime);
+    } catch (Exception e) {
+      e.printStackTrace();
+      promise.reject("ERR_APPLICATION", "Unable to get last update time of this application.", e);
+    }
+  }
+
+
+  @ExpoMethod
+  public void getInstallReferrerAsync(final Promise promise) {
+
+    final StringBuilder installReferrer = new StringBuilder();
+
+    final InstallReferrerClient referrerClient;
+    referrerClient = InstallReferrerClient.newBuilder(mContext).build();
+    referrerClient.startConnection(new InstallReferrerStateListener() {
+      @Override
+      public void onInstallReferrerSetupFinished(int responseCode) {
+        switch (responseCode) {
+          case InstallReferrerClient.InstallReferrerResponse.OK:
+            // Connection established
+            Log.d("INSTALL_REFERRER", "connection established");
+            try {
+              ReferrerDetails response = referrerClient.getInstallReferrer();
+              installReferrer.append(response.getInstallReferrer());
+              response.getReferrerClickTimestampSeconds();
+              response.getInstallBeginTimestampSeconds();
+              promise.resolve(installReferrer.toString());
+              referrerClient.endConnection();
+            } catch (RemoteException e) {
+              e.printStackTrace();
+              promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "Binding to play store remote invocation error.", e);
+              referrerClient.endConnection();
+            }
+            break;
+          case InstallReferrerClient.InstallReferrerResponse.FEATURE_NOT_SUPPORTED:
+            // API not available on the current Play Store app
+            Log.d("INSTALL_REFERRER", "feature not supported");
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "This feature not supported");
+            referrerClient.endConnection();
+            break;
+          case InstallReferrerClient.InstallReferrerResponse.SERVICE_UNAVAILABLE:
+            // Connection could not be established
+            Log.d("INSTALL_REFERRER", "connection could not be established");
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "Connection could not be established");
+            referrerClient.endConnection();
+            break;
+          default:
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "Connection could not be established");
+            referrerClient.endConnection();
```

Comment:
Seems like we always want to end the connection -- better to do it in a finally block